src/coreclr/vm/proftoeeinterfaceimpl.cpp

```diff
@@ -9141,6 +9141,15 @@ HRESULT ProfToEEInterfaceImpl::GetObjectGeneration(ObjectID objectId,
 
     IGCHeap *hp = GCHeapUtilities::GetGCHeap();
 
+    if (hp->IsInFrozenSegment((Object*)objectId))
+    {
+        range->generation = (COR_PRF_GC_GENERATION)-1;
+        range->rangeStart = 0;
+        range->rangeLength = 0;
+        range->rangeLengthReserved = 0;
+        return CORPROF_E_NOT_GC_OBJECT;
```

Comment:
Returning error here makes this more breaking than necessary. Why can't we just return `S_OK`?

Reply:
cc @noahfalk

Reply:
Returning -1 seemed like it was going to be a breaking change no matter what so I figured we may as well make it more obvious rather than try to minimize it and increase risk that tools break more subtly. However we don't yet have feedback from profiler authors one way or another and this behavior is just a best guess on my part. If you think we'd be better off defaulting to keeping S_OK @jkotas I don't have a strong opinion on it.

Reply:
Traditionally GC generations imply that genX can be collected without collecting generations X+.

`-1` makes impression that any level of GC will collect frozen, while in fact it is the opposite - any generation can be collected without touching frozen. 

That seems to imply that frozen objects logically belong to generation > MAX_GEN. 
Perhaps `3` or MAXINT could work better that `-1` ?

Reply:
> If you think we'd be better off defaulting to keeping S_OK

It looks odd to me to return failure error code and still try to return meaningful values. I see that there are other profiler APIs that work like that, so it should be fine to return error code. You can ignore my feedback.

Reply:
I kind of agree with @VSadov that e.g. INT_MAX value would be safer, e.g. I already had to apply a hack in code because GC/VM usages of `WhichGeneration` participate in comparisons so -1 will trigger a different path than previously `max_generation` did.

Reply:
> Perhaps 3 or MAXINT could work better that -1

I have no particular preference between MAXINT and -1. I don't anticipate its going to make a difference myself but I also have no concern changing it. 3 I would not use - in the right context 3 is used as a legitimate generation number and the goal is that nobody should treat this value as if it is a generation.

Reply:
There are not many purposes of `WhichGeneration` other than to answer questions like “can this object be collected before that one?” or “can this object survive GC X?”.  

Mapping immortal objects to gen `-1` feels like not the most convenient/intuitive.

Reply:
I tried INT_MAX and there was only one place that I had to change (some NativeAOT test expecting frozen objects to be in Gen2). Also, `GC.Collect(GC.GetGeneration..` is not breaking now as Jan highlighted.
For -1 it becomes a little bit more complicated.

So any strong opinions for keeping -1 ?

Reply:
I think ` INT_MAX` is the correct value. Even with possible future changes, I do not see anything collecting less often than frozen.